Haptics/app/src/main/java/com/example/android/haptics/samples/ui/resist/ResistRoute.kt

```diff
@@ -0,0 +1,335 @@
+        targetValue = if (isDragging) size else START_SIZE,
+        animationSpec = animationSpecDp
+    )
+    val strokeWidthAnimated by animateDpAsState(
+        targetValue = if (isDragging) strokeWidth else START_STROKE_WIDTH,
+        animationSpec = animationSpecDp
+    )
+    val rotationAnimated by animateFloatAsState(
+        targetValue = if (isDragging) rotation else START_INITIAL_ROTATION,
+        animationSpec = animationSpecFloat
+    )
+    val offsetYAnimated by animateDpAsState(
+        targetValue = if (isDragging) offsetY else START_Y_OFFSET,
+        animationSpec = animationSpecDp
+    )
+
+    // The interval and the intensity of the vibrations.
+    var vibrationInterval by remember { mutableStateOf(TICK_INTERVAL_MAX_MS) }
+    var vibrationIntensity by remember { mutableStateOf(TICK_INTENSITY_MIN) }
+
+    // Use an interpolator to simulate resistance as the user approaches end of their drag.
+    val interpolator = DecelerateInterpolator()
+
+    // Use the drag distance to calculate the indicator state and the vibration intensity.
+    if (dragOffset <= 0) {
+        // Initial state of the indicator.
+        offsetY = START_Y_OFFSET
+        size = START_SIZE
+        strokeWidth = START_STROKE_WIDTH
+        rotation = START_INITIAL_ROTATION
+
+        vibrationInterval = TICK_INTERVAL_MAX_MS
+        vibrationIntensity = TICK_INTENSITY_MIN
+    } else if (dragOffset < maxDragOffset) {
+        // User is dragging between start and max drag distance.
+        val percentDragInterpolated = interpolator.getInterpolation(dragOffset / maxDragOffset)
+        val relativeInterval =
+            ((1f - percentDragInterpolated) * (TICK_INTERVAL_MAX_MS - TICK_INTERVAL_MIN_MS)).toLong()
+        val relativeIntensity =
+            percentDragInterpolated * (TICK_INTENSITY_MAX - TICK_INTENSITY_MIN)
+        val additionalSizeBasedOnDrag = Dp(percentDragInterpolated * (TARGET_SIZE.value - START_SIZE.value))
+        val additionalStrokeWidthBasedOnDrag =
+            Dp((percentDragInterpolated * (TARGET_STROKE_SIZE.value - START_STROKE_WIDTH.value)))
+        val additionalRotation = percentDragInterpolated * TARGET_ROTATION
+        offsetY = Dp(percentDragInterpolated * TARGET_Y_OFFSET.value)
+        size = START_SIZE + additionalSizeBasedOnDrag
+        strokeWidth = START_STROKE_WIDTH + additionalStrokeWidthBasedOnDrag
+        rotation = additionalRotation + START_INITIAL_ROTATION
+
+        vibrationInterval = TICK_INTERVAL_MIN_MS + relativeInterval
+        vibrationIntensity = TICK_INTENSITY_MIN + relativeIntensity
+    } else {
+        // Max size, offset, and vibration intensity/frequency.
+        offsetY = TARGET_Y_OFFSET
+        size = TARGET_SIZE
+        strokeWidth = TARGET_STROKE_SIZE
+        rotation = START_INITIAL_ROTATION + TARGET_ROTATION
+
+        vibrationInterval = TICK_INTERVAL_MIN_MS
+        vibrationIntensity = TICK_INTENSITY_MAX
+    }
+
+    LaunchedEffect(Unit) {
```

Comment:
I just saw the documentation for LaunchedEffect, and it seems like if you simply calls this from a `if (isDragging) { ... }` block then it should only launch the effect when it becomes true and cancel it when it becomes false:

https://developer.android.com/jetpack/compose/side-effects#launchedeffect

Could you please give it a try, to avoid keeping this running while the user is not interacting with the indicator?

Reply:
Yeah I see what you mean. Use the variable to determine whether the LaunchedEffect is composed